torch/distributed/_spmd/iter_graph_module.py

```diff
@@ -0,0 +1,393 @@
+    cleanup graph. The 3 graphs should be identical copies of a ``fx.Graph``.
+
+    IterGraph subclass fx.Graph to override the necessary APIs that will be used
+    when constructing a optimization, e.g., communication fusion. IterGraph also
+    provides APIs that originally belong to fx.Node and all these APIs will have
+    ``node_`` prefix. For example, ``IterGraph.node_prepend`` is the equivalance
+    of ``fx.Node.prepend``. Note that all the optimizations must be constructed
+    using these APIs.
+    """
+
+    def __init__(
+        self,
+        orig_graph: fx.Graph,
+        setup_graph: fx.Graph,
+        cleanup_graph: fx.Graph,
+        owning_module: Optional[fx.GraphModule] = None,
+        tracer_cls: Optional[Type["fx.Tracer"]] = None,
+        tracer_extras: Optional[Dict[str, Any]] = None,
+    ):
+        super().__init__(owning_module, tracer_cls, tracer_extras)
+
+        output_vals = self.graph_copy(orig_graph, {}, return_output_node=True)
+        self._codegen = copy.deepcopy(orig_graph._codegen)
+        assert isinstance(output_vals, tuple)
+        output_val, old_output_val = output_vals
+        self.output(output_val, type_expr=getattr(old_output_val, "type", None))
+
+        self.setup_graph = setup_graph
+        self.cleanup_graph = cleanup_graph
+        self._all_graphs: Tuple[fx.Graph, ...] = (
+            self,
+            self.setup_graph,
+            self.cleanup_graph,
+        )
+
+        self._setup_mapping: Dict[fx.Node, fx.Node] = {}
+        self._cleanup_mapping: Dict[fx.Node, fx.Node] = {}
+
+        for node, setup_node, cleanup_node in zip(
+            self.nodes, self.setup_graph.nodes, self.cleanup_graph.nodes
+        ):
+            self._setup_mapping[node] = setup_node
+            self._cleanup_mapping[node] = cleanup_node
+
+        self.num_extra_output = 0
+
+    def _lookup_node(self, node: fx.Node, graph: fx.Graph) -> Optional[fx.Node]:
+        if graph == self.setup_graph:
+            return self._setup_mapping.get(node, None)
+        elif graph == self.cleanup_graph:
+            return self._cleanup_mapping.get(node, None)
+        return node
+
+    def _insert_context(self, func: str, node: fx.Node):
+        with ExitStack() as stack:
+            for graph in self._all_graphs:
+                if node:
+                    actual_node = self._lookup_node(node, graph)
+                    assert actual_node is not None, "Cannot handle None case now."
+                else:
+                    actual_node = node
+                stack.enter_context(getattr(graph, func)(actual_node))
+            yield
+
+    @contextmanager
+    def inserting_after(self, node):
+        return self._insert_context("inserting_after", node)
+
+    @contextmanager
+    def inserting_before(self, node):
+        return self._insert_context("inserting_before", node)
+
+    @staticmethod
+    def _is_sese_graph(nodes: List[fx.Node], graph: fx.Graph) -> bool:
```

Comment:
Does this mean that the output of this graph cannot be the input of two following subgraphs?

Reply:
It can but there can only be one output. However, I found that this is to restrictive. Will change this in the next few PRs.